spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageResourceObject.java

```diff
@@ -121,20 +121,32 @@ public Blob getGoogleStorageObject() throws IOException {
 	}
 
 	/**
-	 * Creates a signed URL to an object if it exists. This method will fail if this storage resource
-	 * was not created using service account credentials.
+	 * Creates a signed URL to an object.
+	 *
+	 * <p>If an object doesn't exist and {@code createBlobIfNotExists} is {@code true}, then the
+	 * object is created in GCS.
+	 *
+	 * <p>This method will fail if this storage resource was not created using service account
+	 * credentials.
 	 * @param timeUnit the time unit used to determine how long the URL is valid.
 	 * @param timePeriods the number of periods to determine how long the URL is valid.
 	 * @return the URL if the object exists, and null if it does not.
 	 * @throws IOException
 	 */
 	public URL createSignedUrl(TimeUnit timeUnit, long timePeriods) throws IOException {
 		Blob blob = this.getGoogleStorageObject();
+
 		if (blob == null) {
-			return null;
+			if (!this.createBlobIfNotExists) {
```

Comment:
Why not just create `BlobInfo` without the blob using:
```java
BlobInfo.newBuilder(getBlobId()).build()
```

Does the creation of the signed URL actually required the `Blob` to exist in GCS?

I don't think `createSignedUrl` should create the blob. It seems like an unexpected side effect.

Reply:
@meltsufin I was thinking originally that made sense, because how would you have a link to download a nonexistent file right? But the signed URLs allow you to write to the object, too, so an empty blob needs to be created if the signed URL is going to be written to.

Reply:
I think you can send PUT requests to that URL, so it wouldn't be via Spring and our methods here.

Reply:
The issue here is when you create a blob programatically, don't write to it and then try to get the signed URL for that empty blob. This operation should succeed but is currently failing.
I share your concerns that creating a blob is an unexpected side effect of a method called `createSignedUrl()`, so we'll probably need to expose the blob creation mechanism and call it from `createRelative`..

Reply:
I just tested this. You can create a signed URL to a non-existent object, but when you try to do a PUT to it, you'll get an error like "Anonymous caller does not have storage.objects.create access to [file]".

I think we should allow creation of the signed URL for non-existent objects, but output a warning from the method when that happens.

Reply:
Why not just let the auto-create setting govern the behavior? It's very reasonable for a user to assume a writeable signed URL is given if he has the auto-create setting on.

Reply:
Imagine a sequence of calls like this:
```
file1 = new GoogleStorageResourceObject(storage, "file1", true)
file1.exists() // (false)
file1.createSignedUrl()
file1.exists() // (true)
```

You don't think this a bit surprising?

Reply:
@meltsufin I would only find that surprising if I also found this surprising:

```
file1 = new GoogleStorageResourceObject(storage, "file1", true)
file1.exists() // (false)
file1.getOutputStream()
file1.exists() // (true)
```

If you keep in mind that a signed URL is as writeable as an output stream then neither should be surprising.